Initial Support for Remote API v1.18

diptanu opened the pull request:
Added support for the following -

a. Labels
b. Mac Address
b. CPU, CPU Set, Memory, Swap, Parent cgroup

diptanu commented:
@davidxia @rohansingh  The test failed because we couldn't find Docker 1.6.2 I guess. Not sure how the test environment is setup. Any pointers?

davidxia commented:
We use circleci which last I checked only supports up to 1.5. Add an
'assumes'?

On Thursday, June 4, 2015, Diptanu Choudhury <notifications@github.com>
wrote:

> @davidxia <https://github.com/davidxia> @rohansingh
> <https://github.com/rohansingh> The test failed because we couldn't find
> Docker 1.6.2 I guess. Not sure how the test environment is setup. Any
> pointers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/190#issuecomment-109029845>
> .
>


-- 
David Xia
Software Engineer
dxia@spotify.com

diptanu commented:
@davidxia @rohansingh  Not sure why rest of the tests are failing. They work locally. Can you guys take a look please? I have tested this out on our servers, works fine now. cpu.shares is getting set appropriately.  What are the next steps here?

davidxia commented on src/main/java/com/spotify/docker/client/messages/ContainerConfig.java:
```diff
@@ -629,6 +668,33 @@ public Builder onBuild(final String... onBuild) {
       return onBuild;
     }
 
+    public Builder labels(Map<String, String> labels) {
```
Add a `final` here for consistency?

davidxia commented on src/main/java/com/spotify/docker/client/messages/HostConfig.java:
```diff
@@ -46,6 +46,11 @@
+  @JsonProperty("Memory") private Long memory;
+  @JsonProperty("MemorySwap") private Long memorySwap;
+  @JsonProperty("CpuShares") private Long cpuShares;
+  @JsonProperty("CpusetCpus") private String cpuset;
```
Any reason to use `cpuset` instead of `cpisetCpus`? The latter would be more consistent with the docker API.

davidxia commented on src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
```diff
@@ -935,25 +938,31 @@ public void testInspectContainerWithSecurityOpts() throws Exception {
   @Test
   public void testContainerWithHostConfig() throws Exception {
+    assumeTrue("Docker API should be at least v1.18 to support Container Creation with HostConfig, got "
+            + sut.version().apiVersion(), versionCompare(sut.version().apiVersion(), "1.18") >= 0);
```
nitpick: line up the `+` with `"` above.

davidxia commented on src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
```diff
@@ -935,25 +938,31 @@ public void testInspectContainerWithSecurityOpts() throws Exception {
 
   @Test
   public void testContainerWithHostConfig() throws Exception {
+    assumeTrue("Docker API should be at least v1.18 to support Container Creation with HostConfig, got "
```
style nitpick: don't exceed 100 chars on a line

davidxia commented on src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
```diff
@@ -935,25 +938,31 @@ public void testInspectContainerWithSecurityOpts() throws Exception {
 
   @Test
   public void testContainerWithHostConfig() throws Exception {
```
Maybe we should add an extra test specifically for container creation where the HostConfig is in the ContainerConfig so the assumeTrue doesn't ignore this test for docker API versions less than 1.18.

davidxia replied:
For reference, Docker API 1.15 is the last version where `/containers/(id)/start` took a HostConfig JSON in the POST data. https://docs.docker.com/reference/api/docker_remote_api_v1.15/#start-a-container

diptanu replied:
@davidxia Yes, I noticed that. So are you guys willing to drop support for 1.15 entirely? I am happy to update the PR to make the API conform to only API version 1.18

davidxia replied:
I'd prefer to keep this PR small. So maybe include an assumeTrue for this test for docker API <= 1.15 and rename it to something like `testContainerWithHostConfigOldApi`. Then add another `testContainerWithHostConfig` with your changes above?

@rohansingh thoughts?

diptanu commented:
@davidxia Addressed your review comments. Waiting to hear if you guys want to drop support for versions less than 1.15 completely since CPU shares, CPU Set etc aren't in ContainerConfig anymore after v1.15 and the container start api doesn't take HostConfig anymore. I can update the PR to reflect the new API

rohansingh commented:
@diptanu Just discussed it and we're OK with dropping support for older versions. We need to figure out how to annotate docker-client version <-> Docker version compatibility, but that's a separate issue.

diptanu commented:
@rohansingh Thanks, let me update this PR now.

diptanu commented:
@rohansingh @davidxia I removed the start method in DockerClient which took HostConfig. Please let me know if you guys have more comments?

I would like to get this PR merged and then work on the remaining items which would add support for more properties in the HostConfig like privileges, log driver configuration, RestartPolicy, ReadonlyRootfs, etc.

diptanu commented:
CircleCI shows Build Succeded and all tests are passing not sure why the PR is not updated with the success callback from Circle CI.

rohansingh commented:
Green light from CircleCI now. @davidxia I'll leave pushing the button up to you since you reviewed.

davidxia commented:
@diptanu Amazing! Thanks for this clean commit and the fast replies.

diptanu commented:
@davidxia @rohansingh Thanks! More commits coming your way to add support for rest of the properties in HostConfig
